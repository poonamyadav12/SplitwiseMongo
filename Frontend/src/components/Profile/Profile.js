import React, { Component, useState } from 'react';
import { Button, Card, Col, Container, Form, ListGroup, Row } from 'react-bootstrap';
import { Typeahead } from 'react-bootstrap-typeahead';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import TimezoneSelect from 'react-timezone-select';
import '../../App.css';
import { userActions } from '../../_actions';
import { getDefaultUserImage } from '../../_constants/avatar';
import { CURRENCY } from '../../_helper/money';
import { UploadImage } from '../Image/UploadImage';
import { GrEdit } from 'react-icons/gr';
import PhoneInput from 'react-phone-input-2'
import 'react-phone-input-2/lib/style.css'

//Define a Login Component
class Profile extends Component {
  //call the constructor method
  constructor(props) {
    //Call the constrictor of Super class i.e The Component
    super(props);
    //maintain the state required for this component
    this.state = {
      firstName: this.props.user?.first_name,
      lastName: this.props.user?.last_name,
      username: this.props.user?.email,
      password: null,
      new_password: null,
      timeZone: this.props.user?.time_zone,
      currency: this.props.user?.default_currency,
      imageUrl: this.props.user?.avatar,
      editMode: false,
      phoneNumber: this.props.user?.phone_number
    };

    //Bind the handlers to this class
    this.firstNameChangeHandler = this.firstNameChangeHandler.bind(this);
    this.lastNameChangeHandler = this.lastNameChangeHandler.bind(this);
    this.usernameChangeHandler = this.usernameChangeHandler.bind(this);
    this.currentPasswordChangehandler = this.currentPasswordChangehandler.bind(this);
    this.newPasswordChangeHandler = this.newPasswordChangeHandler.bind(this);
    this.submitSignup = this.submitSignup.bind(this);
    this.timezoneHandler = this.timezoneHandler.bind(this);
    this.currencyHandler = this.currencyHandler.bind(this);
    this.avatarChangeHandler = this.avatarChangeHandler.bind(this);
    this.setPhoneNumber = this.setPhoneNumber.bind(this);

  }
  //Call the Will Mount to set the auth Flag to false
  componentWillMount() {
    this.setState({
      authFlag: false
    })
  }
  avatarChangeHandler = (imageUrl) => {
    this.setState({ imageUrl });
  }

  //name change handler to update state variable with the text entered by the user
  firstNameChangeHandler = (e) => {
    this.setState({
      firstName: e.target.value
    })
  }
  //name change handler to update state variable with the text entered by the user
  lastNameChangeHandler = (e) => {
    this.setState({
      lastName: e.target.value
    })
  }
  //username change handler to update state variable with the text entered by the user
  usernameChangeHandler = (e) => {
    this.setState({
      username: e.target.value
    })
  }
  //password change handler to update state variable with the text entered by the user
  currentPasswordChangehandler = (e) => {
    this.setState({
      password: e.target.value
    })
  }
  //password change handler to update state variable with the text entered by the user
  newPasswordChangeHandler = (e) => {
    this.setState({
      new_password: e.target.value
    })
  }
  //timezone change handler
  timezoneHandler = (value) => {
    this.setState({
      timeZone: value.value
    })
  }
  //timezone change handler
  currencyHandler = (value) => {
    this.setState({
      currency: value[0].name
    })
  }
  setPhoneNumber = (value) => {
    this.setState({
      phoneNumber: value
    })
  }
  //submit Login handler to send a request to the node backend
  submitSignup = (e) => {
    //var headers = new Headers();
    //prevent page from refresh
    e.preventDefault();
    const data = {
      user: {
        first_name: this.state.firstName,
        last_name: this.state.lastName,
        email: this.state.username,
        default_currency: this.state.currency,
        password: this.props.user.password,
        time_zone: this.state.timeZone,
        avatar: this.state.imageUrl,
        phone_number: this.state.phoneNumber,
      }
    }

    if (this.state.new_password) {
      data.user.password = this.state.password;
      data.user.new_password = this.state.new_password;
    }

    this.props.update(data);
    this.setState({ editMode: false });
  }

  setEditMode() {
    this.setState({ editMode: true });
  }

  render() {
    console.log('Here ' + this.props.user);
    return (
      <>
        {!this.props.user && <Redirect to='/login' />}
        {this.props.user &&
          <Container>
            <Row style={{ alignItems: 'center' }}>
              <Col lg={3} style={{ 'width': '20%', marginTop: '3%' }}>
                <img style={{ width: '17rem' }} src="https://assets.splitwise.com/assets/core/logo-square-65a6124237868b1d2ce2f5db2ab0b7c777e2348b797626816400534116ae22d7.svg" alt="Logo" />
              </Col>
              <Col lg={3}>
                <UploadImage label="Your avatar" defaultImage={this.state.imageUrl || getDefaultUserImage()} onChange={this.avatarChangeHandler}></UploadImage>
              </Col>
              <Col lg={6}>
                <Card style={{ width: '35rem' }}>
                  <Form>
                    <EditableField parentReloaded={this.state.reloaded} setParentEditMode={this.setEditMode.bind(this)} controlId="formFirstName" label="First name" value={this.state.firstName} onChange={this.firstNameChangeHandler} />
                    <EditableField parentReloaded={this.state.reloaded} setParentEditMode={this.setEditMode.bind(this)} controlId="formLastName" label="Last name" value={this.state.lastName} onChange={this.lastNameChangeHandler} />
                    <EditablePhoneNumberField parentReloaded={this.state.reloaded} setParentEditMode={this.setEditMode.bind(this)} controlId="formPhoneNumber" label="Phone Number" value={this.state.phoneNumber} onChange={this.setPhoneNumber} />
                    <Form.Group controlId="formBasicEmail">
                      <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'baseline' }}>
                        <Form.Label>Email address</Form.Label>
                        <Form.Label>{this.props.user.email}</Form.Label>
                      </div>
                    </Form.Group>
                    <EditablePasswordField parentReloaded={this.state.reloaded} setParentEditMode={this.setEditMode.bind(this)} controlId="formLastName" currentPassword={this.state.password} newPassword={this.state.new_password} onChangeCurrent={this.currentPasswordChangehandler} onChangeNew={this.newPasswordChangeHandler} />
                    <Form.Group>
                      <Form.Label>Currency</Form.Label>
                      <Typeahead
                        id="currency"
                        labelKey="name"
                        onChange={this.currencyHandler}
                        options={CURRENCY}
                        placeholder="Choose a currency..."
                        defaultInputValue={this.state.currency}
                        renderMenuItemChildren={(option, props) => (
                          <ListGroup.Item key={option.name}>
                            <span>{option.name}</span>{' '}
                          </ListGroup.Item>
                        )}
                      />
                    </Form.Group>
                    <Form.Group>
                      <Form.Label>Time Zone</Form.Label>
                      <TimezoneSelect
                        value={'America/Los Angeles'}
                        onChange={this.timezoneHandler}
                        defaultInputValue={this.state.timeZone}
                        timezones={{
                          'America/Los_Angeles': 'Los Angeles',
                          'Europe/Berlin': 'Frankfurt',
                          'Asia/Kuwait': 'Kuwait',
                          'Asia/Bahrain': 'Bahrain',
                          'Canada/Mountain': 'Canada Mountain',
                          'Canada/Pacific': 'Canada Pacific',
                          'Asia/Kolkata': 'Indian Standard Time',
                          //    ...i18nTimezones,
                        }}
                      />
                    </Form.Group>
                    <Button variant="success" type="submit" onClick={this.submitSignup}>
                      Submit</Button>
                  </Form>
                </Card>
              </Col>
            </Row>
          </Container>}
      </>
    )
  }
}

const EditableField = (props) => {
  const [editMode, setEditMode] = useState(props.editMode);
  if (editMode) {
    return <Form.Group controlId={`${props.controlId}Edit`}>
      <Form.Label>{props.label}</Form.Label>
      <Form.Control type="text" value={props.value || ''} onChange={props.onChange} />
    </Form.Group>;
  }
  return <Form.Group controlId={`${props.controlId}View`}>
    <Form.Label>{props.label}</Form.Label>
    <div style={{ display: 'flex', alignItems: 'baseline' }}>
      <Form.Label>{props.value || 'none'}</Form.Label> &nbsp; <GrEdit onClick={() => { setEditMode(true); props.setParentEditMode(true); }} />
    </div>
  </Form.Group>;
}

const EditablePasswordField = (props) => {
  const [editMode, setEditMode] = useState(props.editMode);
  if (editMode) {
    return <><Form.Group controlId={`${props.controlId}CurrentEdit`}>
      <Form.Label>Current Password</Form.Label>
      <Form.Control type="password" value={props.currentPassword || ''} onChange={props.onChangeCurrent} />
    </Form.Group>
      <Form.Group controlId={`${props.controlId}NewEdit`}>
        <Form.Label>New Password</Form.Label>
        <Form.Control type="password" value={props.newPassword || ''} onChange={props.onChangeNew} />
      </Form.Group>
    </>;
  }
  return <Form.Group controlId={`${props.controlId}View`}>
    <Form.Label>{props.label}</Form.Label>
    <div style={{ display: 'flex', alignItems: 'baseline' }}>
      <Form.Label>••••••••••</Form.Label> &nbsp; <GrEdit onClick={() => { setEditMode(true); props.setParentEditMode(true); }} />
    </div>
  </Form.Group>;
}
const EditablePhoneNumberField = (props) => {
  const [editMode, setEditMode] = useState(props.editMode);
  if (editMode) {
    return <><Form.Group controlId={`${props.controlId}Edit`}>
      <Form.Label>{props.label}</Form.Label>
      <PhoneInput
        placeholder={props.value || ''}
        value={props.value || ''}
        onChange={props.onChange} />
    </Form.Group>
    </>;
  }
  return <Form.Group controlId={`${props.controlId}View`}>
    <Form.Label>{props.label}</Form.Label>
    <div style={{ display: 'flex', alignItems: 'baseline' }}>
      <Form.Label>{props.value || 'none'}</Form.Label> &nbsp; <GrEdit onClick={() => { setEditMode(true); props.setParentEditMode(true); }} />
    </div>
  </Form.Group>;
}

function mapState(state) {
  const { user } = state.authentication;
  const { updating } = state.update;
  return { updating, user };
}

const actionCreators = {
  update: userActions.update
}

const connectedProfilePage = connect(mapState, actionCreators)(Profile);
export { connectedProfilePage as Profile, Profile as ProfileForTest };