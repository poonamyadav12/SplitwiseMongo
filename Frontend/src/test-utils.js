// test-utils.js
import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
// Import your own reducer
import rootReducer from '../src/reducers';
import { createLogger } from 'redux-logger';
import { createStore, applyMiddleware, compose } from 'redux';
import thunkMiddleware from 'redux-thunk';

jest.mock('react-router-dom'); // <- Use this line
jest.mock('react-avatar-editor');
jest.mock('axios');

const loggerMiddleware = createLogger();
const composeEnhancers = (typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

function render(
  ui,
  {
    initialState,
    store = createStore(
      rootReducer,
      composeEnhancers(
        applyMiddleware(
          thunkMiddleware,
          loggerMiddleware
        ))
    ),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }