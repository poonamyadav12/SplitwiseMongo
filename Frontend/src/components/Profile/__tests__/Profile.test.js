import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import mockAxios from "axios";
import { setupServer } from 'msw/node';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { fireEvent, render, screen } from '../../../test-utils';
import { ProfileForTest } from '../Profile';
import renderer from 'react-test-renderer';


const server = setupServer()

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

let updateValue = {};
const user = {
  "email": "abc@gmail.com",
  "first_name": "Michael",
  "last_name": "Clarke",
  "password": "Abc@1234",
  "time_zone": "Asia/Kolkata",
  "default_currency": "USD",
  "avatar": "www.example.com",
};
const update = (user) => { updateValue = user };

test('displays profile page and makes API call', async () => {

  mockAxios.post.mockImplementationOnce(() =>
    Promise.resolve({
      data: user
    })
  );

  render(<BrowserRouter><ProfileForTest user={user} update={update} /></BrowserRouter>)

  const firstName = screen.getByText('Michael');
  const lastName = screen.getByText('Clarke');
  const email = screen.getByText('abc@gmail.com');

  fireEvent.click(screen.getByText('Submit'));

  expect(updateValue).toEqual({ user });
});

test('renders correctly', () => {
  const tree = renderer
    .create(<ProfileForTest user={user} update={update}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
