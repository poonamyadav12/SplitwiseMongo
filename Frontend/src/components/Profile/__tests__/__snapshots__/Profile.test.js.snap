// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <div
      className="col-lg-3"
      style={
        Object {
          "marginTop": "3%",
          "width": "20%",
        }
      }
    >
      <img
        alt="Logo"
        src="https://assets.splitwise.com/assets/core/logo-square-65a6124237868b1d2ce2f5db2ab0b7c777e2348b797626816400534116ae22d7.svg"
        style={
          Object {
            "width": "17rem",
          }
        }
      />
    </div>
    <div
      className="col-lg-3"
    >
      <div
        className="col"
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="imageUplaod"
          >
            <h5>
              Your avatar
            </h5>
          </label>
          <div
            onBlur={[Function]}
            onClick={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
          >
            <input
              autoComplete="off"
              multiple={true}
              onChange={[Function]}
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              tabIndex={-1}
              type="file"
            />
          </div>
          <h5>
            Drop images or click to upload
          </h5>
        </div>
      </div>
    </div>
    <div
      className="col-lg-6"
    >
      <div
        className="card"
        style={
          Object {
            "width": "35rem",
          }
        }
      >
        <form
          className=""
        >
          <div
            className="form-group"
          >
            <label
              className="form-label"
              htmlFor="formFirstNameView"
            >
              First name
            </label>
            <div
              style={
                Object {
                  "alignItems": "baseline",
                  "display": "flex",
                }
              }
            >
              <label
                className="form-label"
                htmlFor="formFirstNameView"
              >
                Michael
              </label>
                 
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14,4 L20,10 L14,4 Z M22.2942268,5.29422684 C22.6840146,5.68401459 22.6812861,6.3187139 22.2864907,6.71350932 L9,20 L2,22 L4,15 L17.2864907,1.71350932 C17.680551,1.319449 18.3127724,1.31277239 18.7057732,1.70577316 L22.2942268,5.29422684 Z M3,19 L5,21 M7,17 L15,9"
                  fill="none"
                  stroke="#000"
                  strokeWidth="2"
                />
              </svg>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
              htmlFor="formLastNameView"
            >
              Last name
            </label>
            <div
              style={
                Object {
                  "alignItems": "baseline",
                  "display": "flex",
                }
              }
            >
              <label
                className="form-label"
                htmlFor="formLastNameView"
              >
                Clarke
              </label>
                 
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14,4 L20,10 L14,4 Z M22.2942268,5.29422684 C22.6840146,5.68401459 22.6812861,6.3187139 22.2864907,6.71350932 L9,20 L2,22 L4,15 L17.2864907,1.71350932 C17.680551,1.319449 18.3127724,1.31277239 18.7057732,1.70577316 L22.2942268,5.29422684 Z M3,19 L5,21 M7,17 L15,9"
                  fill="none"
                  stroke="#000"
                  strokeWidth="2"
                />
              </svg>
            </div>
          </div>
          <div
            className="form-group"
          >
            <div
              style={
                Object {
                  "alignItems": "baseline",
                  "display": "flex",
                  "flexDirection": "column",
                }
              }
            >
              <label
                className="form-label"
                htmlFor="formBasicEmail"
              >
                Email address
              </label>
              <label
                className="form-label"
                htmlFor="formBasicEmail"
              >
                abc@gmail.com
              </label>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
              htmlFor="formLastNameView"
            />
            <div
              style={
                Object {
                  "alignItems": "baseline",
                  "display": "flex",
                }
              }
            >
              <label
                className="form-label"
                htmlFor="formLastNameView"
              >
                ••••••••••
              </label>
                 
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14,4 L20,10 L14,4 Z M22.2942268,5.29422684 C22.6840146,5.68401459 22.6812861,6.3187139 22.2864907,6.71350932 L9,20 L2,22 L4,15 L17.2864907,1.71350932 C17.680551,1.319449 18.3127724,1.31277239 18.7057732,1.70577316 L22.2942268,5.29422684 Z M3,19 L5,21 M7,17 L15,9"
                  fill="none"
                  stroke="#000"
                  strokeWidth="2"
                />
              </svg>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Currency
            </label>
            <div
              className="rbt"
              style={
                Object {
                  "outline": "none",
                  "position": "relative",
                }
              }
              tabIndex={-1}
            >
              <div
                style={
                  Object {
                    "display": "flex",
                    "flex": 1,
                    "height": "100%",
                    "position": "relative",
                  }
                }
              >
                <input
                  aria-autocomplete="both"
                  aria-expanded={false}
                  aria-haspopup="listbox"
                  autoComplete="off"
                  className="rbt-input-main form-control rbt-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Choose a currency..."
                  role="combobox"
                  type="text"
                  value="USD"
                />
                <input
                  aria-hidden={true}
                  className="rbt-input-hint"
                  readOnly={true}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderColor": "transparent",
                      "boxShadow": "none",
                      "color": "rgba(0, 0, 0, 0.35)",
                      "left": 0,
                      "pointerEvents": "none",
                      "position": "absolute",
                      "top": 0,
                      "width": "100%",
                    }
                  }
                  tabIndex={-1}
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="form-label"
            >
              Time Zone
            </label>
            <div
              className=" css-2b097c-container"
              onKeyDown={[Function]}
            >
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                className="css-1f43avz-a11yText-A11yText"
              />
              <div
                className=" css-yk16xz-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-g1d714-ValueContainer"
                >
                  <div
                    className="css-b8ldur-Input"
                  >
                    <div
                      className=""
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-autocomplete="list"
                        autoCapitalize="none"
                        autoComplete="off"
                        autoCorrect="off"
                        disabled={false}
                        id="react-select-3-input"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        spellCheck="false"
                        style={
                          Object {
                            "background": 0,
                            "border": 0,
                            "boxSizing": "content-box",
                            "color": "inherit",
                            "fontSize": "inherit",
                            "label": "input",
                            "opacity": 1,
                            "outline": 0,
                            "padding": 0,
                            "width": "1px",
                          }
                        }
                        tabIndex="0"
                        type="text"
                        value="Asia/Kolkata"
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        Asia/Kolkata
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    className=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-tlfecz-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-tj5bde-Svg"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            className="btn btn-success"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
